File: Plugins/AM_Extraction_PlugIn/clsPeptideProphetWrapper.vb
'*********************************************************************************************************
' Written by Dave Clark for the US Department of Energy 
' Pacific Northwest National Laboratory, Richland, WA
' Copyright 2007, Battelle Memorial Institute
' Created 07/12/2007
'
' Program converted from original version written by J.D. Sandoval, PNNL.
' Conversion performed as part of upgrade to VB.Net 2005, modification for use with manager and broker databases
'
'*********************************************************************************************************

Imports System.IO
Imports AnalysisManagerBase

''' <summary>
''' Calls the PeptideProphetRunner application
''' </summary>
''' <remarks></remarks>
Public Class clsPeptideProphetWrapper

#Region "Constants"
    Public Const MAX_PEPTIDE_PROPHET_RUNTIME_MINUTES As Integer = 120
#End Region

#Region "Module variables"
    Private ReadOnly m_PeptideProphetRunnerLocation As String = String.Empty
	Private m_ErrMsg As String = String.Empty
    Private m_DebugLevel As Short = 1
    Private m_InputFile As String = String.Empty

    Protected WithEvents CmdRunner As clsRunDosProgram

#End Region

    Public Event PeptideProphetRunning(PepProphetStatus As String, PercentComplete As Single)

#Region "Properties"
    Public Property DebugLevel() As Short
        Get
            Return m_DebugLevel
        End Get
        Set(value As Short)
            m_DebugLevel = value
        End Set
    End Property


    Public ReadOnly Property ErrMsg() As String
        Get
            If m_ErrMsg Is Nothing Then
                Return String.Empty
            Else
                Return m_ErrMsg
            End If
        End Get
    End Property

    Public Property InputFile() As String
        Get
            Return m_InputFile
        End Get
        Set(Value As String)
            m_InputFile = Value
        End Set
    End Property

    Public Property Enzyme As String = String.Empty

    Public Property OutputFolderPath As String = String.Empty

#End Region

#Region "Methods"

    Public Sub New(strPeptideProphetRunnerLocation As String)
        m_PeptideProphetRunnerLocation = strPeptideProphetRunnerLocation
    End Sub

    Public Function CallPeptideProphet() As IJobParams.CloseOutType

        Dim CmdStr As String
        Dim ioInputFile As FileInfo
        Dim strPeptideProphetConsoleOutputFilePath As String

        Try
            m_ErrMsg = String.Empty

            ioInputFile = New FileInfo(m_InputFile)
            strPeptideProphetConsoleOutputFilePath = Path.Combine(ioInputFile.DirectoryName, "PeptideProphetConsoleOutput.txt")

            CmdRunner = New clsRunDosProgram(ioInputFile.Directoryname)

            ' verify that program file exists
            If Not File.Exists(m_PeptideProphetRunnerLocation) Then
                m_ErrMsg = "PeptideProphetRunner not found at " & m_PeptideProphetRunnerLocation
                Return IJobParams.CloseOutType.CLOSEOUT_FAILED
            End If

            ' Set up and execute a program runner to run the Peptide Prophet Runner
            CmdStr = ioInputFile.FullName & " " & ioInputFile.DirectoryName & " /T:" & MAX_PEPTIDE_PROPHET_RUNTIME_MINUTES
            If m_DebugLevel >= 2 Then
                clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.DEBUG, m_PeptideProphetRunnerLocation & " " & CmdStr)
            End If

            With CmdRunner
                .CreateNoWindow = True
                .CacheStandardOutput = True
                .EchoOutputToConsole = True

                .WriteConsoleOutputToFile = True
                .ConsoleOutputFilePath = strPeptideProphetConsoleOutputFilePath
            End With

            If Not CmdRunner.RunProgram(m_PeptideProphetRunnerLocation, CmdStr, "PeptideProphetRunner", True) Then
                m_ErrMsg = "Error running PeptideProphetRunner"
                Return IJobParams.CloseOutType.CLOSEOUT_FAILED
            End If

            If CmdRunner.ExitCode <> 0 Then
                m_ErrMsg = "Peptide prophet runner returned a non-zero error code: " & CmdRunner.ExitCode.ToString

                ' Parse the console output file for any lines that contain "Error"
                ' Append them to m_ErrMsg

                Dim ioConsoleOutputFile As FileInfo = New FileInfo(strPeptideProphetConsoleOutputFilePath)
                Dim blnErrorMessageFound As Boolean = False

                If ioConsoleOutputFile.Exists Then
                    Dim srInFile As StreamReader
                    srInFile = New StreamReader(New FileStream(ioConsoleOutputFile.FullName, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))

                    Do While Not srInFile.EndOfStream
                        Dim strLineIn As String
                        strLineIn = srInFile.ReadLine()
                        If Not String.IsNullOrWhiteSpace(strLineIn) Then
                            If strLineIn.ToLower.Contains("error") Then
                                m_ErrMsg &= "; " & m_ErrMsg
                                blnErrorMessageFound = True
                            End If
                        End If
                    Loop
                    srInFile.close()
                End If

                If Not blnErrorMessageFound Then
                    m_ErrMsg &= "; Unknown error message"
                End If

                Return IJobParams.CloseOutType.CLOSEOUT_FAILED
            End If

            RaiseEvent PeptideProphetRunning("Complete", 100)

            Return IJobParams.CloseOutType.CLOSEOUT_SUCCESS

        Catch ex As Exception
            m_ErrMsg = "Exception while running peptide prophet: " & ex.Message & "; " & clsGlobal.GetExceptionStackTrace(ex)
            Return IJobParams.CloseOutType.CLOSEOUT_FAILED
        End Try


    End Function

    ''' <summary>
    ''' Event handler for CmdRunner.LoopWaiting event
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub CmdRunner_LoopWaiting() Handles CmdRunner.LoopWaiting
        Static dtLastStatusUpdate As DateTime = DateTime.UtcNow

        'Update the status (limit the updates to every 5 seconds)
        If DateTime.UtcNow.Subtract(dtLastStatusUpdate).TotalSeconds >= 5 Then
            dtLastStatusUpdate = DateTime.UtcNow
            RaiseEvent PeptideProphetRunning("Running", 50)
        End If

    End Sub

#End Region

End Class

File: Plugins/AM_Extraction_PlugIn/clsPeptideProphetWrapper.cs
//*********************************************************************************************************
// Written by Dave Clark for the US Department of Energy
// Pacific Northwest National Laboratory, Richland, WA
// Copyright 2007, Battelle Memorial Institute
// Created 07/12/2007
//
// Program converted from original version written by J.D. Sandoval, PNNL.
// Conversion performed as part of upgrade to VB.Net 2005, modification for use with manager and broker databases
//
//*********************************************************************************************************

using AnalysisManagerBase;
using PRISM;
using System;
using System.IO;

namespace AnalysisManagerExtractionPlugin
{
    /// <summary>
    /// Calls the PeptideProphetRunner application
    /// </summary>
    /// <remarks></remarks>
    public class clsPeptideProphetWrapper : EventNotifier
    {
        #region "Constants"

        public const int MAX_PEPTIDE_PROPHET_RUNTIME_MINUTES = 120;

        #endregion

        #region "Module variables"

        private readonly string mPeptideProphetRunnerLocation;

        protected clsRunDosProgram mCmdRunner;

        #endregion

        /// <summary>
        /// Even used to report progress
        /// </summary>
        public event PeptideProphetRunningEventHandler PeptideProphetRunning;

        public delegate void PeptideProphetRunningEventHandler(string pepProphetStatus, float percentComplete);

        #region "Properties"

        public short DebugLevel { get; set; } = 1;

        public string ErrMsg { get; private set; } = string.Empty;

        public string InputFile { get; set; } = string.Empty;

        public string Enzyme { get; set; } = string.Empty;

        public string OutputFolderPath { get; set; } = string.Empty;

        #endregion

        #region "Methods"

        public clsPeptideProphetWrapper(string peptideProphetRunnerLocation)
        {
            mPeptideProphetRunnerLocation = peptideProphetRunnerLocation;
        }

        public CloseOutType CallPeptideProphet()
        {
            try
            {
                ErrMsg = string.Empty;

                var inputFile = new FileInfo(InputFile);
                if (inputFile.Directory == null)
                {
                    ReportError("Unable to determine the parent directory of the input file for peptide prophet: " + InputFile);
                    return CloseOutType.CLOSEOUT_FILE_NOT_FOUND;
                }

                var peptideProphetConsoleOutputFilePath = Path.Combine(inputFile.Directory.FullName, "PeptideProphetConsoleOutput.txt");

                // verify that program file exists
                if (!File.Exists(mPeptideProphetRunnerLocation))
                {
                    ReportError("PeptideProphetRunner not found at " + mPeptideProphetRunnerLocation);
                    return CloseOutType.CLOSEOUT_FAILED;
                }

                // Set up and execute a program runner to run the Peptide Prophet Runner
                var arguments =
                    clsGlobal.PossiblyQuotePath(inputFile.FullName) + " " +
                    clsGlobal.PossiblyQuotePath(inputFile.Directory.FullName) +
                    " /T:" + MAX_PEPTIDE_PROPHET_RUNTIME_MINUTES;

                if (DebugLevel >= 2)
                {
                    OnDebugEvent(mPeptideProphetRunnerLocation + " " + arguments);
                }

                mCmdRunner = new clsRunDosProgram(inputFile.Directory.FullName, DebugLevel)
                {
                    CreateNoWindow = true,
                    CacheStandardOutput = true,
                    EchoOutputToConsole = true,
                    WriteConsoleOutputToFile = true,
                    ConsoleOutputFilePath = peptideProphetConsoleOutputFilePath
                };
                mCmdRunner.ErrorEvent += CmdRunner_ErrorEvent;
                mCmdRunner.LoopWaiting += CmdRunner_LoopWaiting;

                if (!mCmdRunner.RunProgram(mPeptideProphetRunnerLocation, arguments, "PeptideProphetRunner", true))
                {
                    ReportError("Error running PeptideProphetRunner");
                    return CloseOutType.CLOSEOUT_FAILED;
                }

                if (mCmdRunner.ExitCode != 0)
                {
                    ReportError("Peptide prophet runner returned a non-zero error code: " + mCmdRunner.ExitCode);

                    // Parse the console output file for any lines that contain "Error"
                    // Append them to mErrMsg

                    var consoleOutputFile = new FileInfo(peptideProphetConsoleOutputFilePath);
                    var errorMessageFound = false;

                    if (consoleOutputFile.Exists)
                    {
                        var reader = new StreamReader(new FileStream(consoleOutputFile.FullName, FileMode.Open, FileAccess.Read, FileShare.ReadWrite));

                        while (!reader.EndOfStream)
                        {
                            var lineIn = reader.ReadLine();
                            if (!string.IsNullOrWhiteSpace(lineIn))
                            {
                                if (lineIn.IndexOf("error", StringComparison.OrdinalIgnoreCase) >= 0)
                                {
                                    ErrMsg += "; " + lineIn;
                                    OnWarningEvent(ErrMsg);
                                    errorMessageFound = true;
                                }
                            }
                        }
                        reader.Close();
                    }

                    if (!errorMessageFound)
                    {
                        ErrMsg += "; Unknown error message";
                        OnWarningEvent("Unknown PeptideProphet error message");
                    }

                    return CloseOutType.CLOSEOUT_FAILED;
                }

                PeptideProphetRunning?.Invoke("Complete", 100);

                return CloseOutType.CLOSEOUT_SUCCESS;
            }
            catch (Exception ex)
            {
                ReportError("Exception while running peptide prophet: " + ex.Message, ex);
                return CloseOutType.CLOSEOUT_FAILED;
            }
        }

        private void ReportError(string message, Exception ex = null)
        {
            ErrMsg = message;
            OnErrorEvent(ErrMsg, ex);
        }

        private void CmdRunner_ErrorEvent(string message, Exception ex)
        {
            ReportError("PeptideProphetRunner: " + ErrMsg);
        }

        private DateTime mLastStatusUpdate = DateTime.MinValue;

        /// <summary>
        /// Event handler for CmdRunner.LoopWaiting event
        /// </summary>
        /// <remarks>Always reports 50% complete</remarks>
        private void CmdRunner_LoopWaiting()
        {
            // Update the status (limit the updates to every 5 seconds)
            if (DateTime.UtcNow.Subtract(mLastStatusUpdate).TotalSeconds >= 5)
            {
                mLastStatusUpdate = DateTime.UtcNow;
                PeptideProphetRunning?.Invoke("Running", 50);
            }
        }

        #endregion
    }
}
